dbsnap_verify/state_doc.py

```diff
@@ -0,0 +1,151 @@
+from os import environ
+
+import json
+
+import time
+
+import boto3
+
+from .rds_funcs import dbsnap_verify_db_id
+
+DB_ID_PREFIX_LEN = 14
+
+# TODO: it feels like `state_doc` should be a dict-like class
+
+
+def now_timestamp():
+    return time.time()
+
+
+def current_state(state_doc):
+    return state_doc["states"][-1]["state"]
+
+
+def get_state_doc_bucket(config):
+    return environ.get("STATE_DOC_BUCKET", config.get("state_doc_bucket", None))
+
+
+def set_state_doc_in_path(state_doc):
+    with open(state_doc["state_doc_path"], 'w') as json_file:
+        json.dump(state_doc, json_file, indent=2)
+
+
+def upload_state_doc(state_doc):
+    state_doc_json = json.dumps(state_doc, indent=2)
+    s3 = boto3.client("s3")
+    s3.put_object(
+        Bucket=get_state_doc_bucket(state_doc),
+        Key=state_doc_s3_key(state_doc["database"]),
+        Body=state_doc_json,
+    )
+
+
+def set_state_doc_in_s3(state_doc):
+    upload_state_doc(state_doc)
+    return state_doc
+
+
+def state_doc_s3_key(database):
+    return "state-doc-{}.json".format(database)
+
+
+def download_state_doc(config):
+    s3 = boto3.client("s3")
+    s3_object = s3.get_object(
+        Bucket=get_state_doc_bucket(config),
+        Key=state_doc_s3_key(config["database"]),
+    )
+    # download state_doc json from s3 and stick it into a string.
+    state_doc_json = s3_object["Body"].read()
+    # turn json into a dict and return it.
+    return json.loads(state_doc_json)
+
+
+def create_state_doc(config):
+    state_doc = config
+    state_doc["tmp_database"] = dbsnap_verify_db_id(state_doc["database"])
+    state_doc["states"] = []
+    return transition_state(state_doc, "wait")
+
+
+def clean_state_doc(state_doc, state_count_to_keep=100):
+    state_doc.pop("tmp_password", None)
+    state_doc["snapshot_verified"] = state_doc.pop("snapshot_verifying", None)
+    trim_index = len(state_doc["states"]) - state_count_to_keep
+    state_doc["states"] = state_doc["states"][trim_index:]
+    return state_doc
+
+
+def get_state_doc_in_s3(config):
+    """get the state_doc in S3 or None."""
+    try:
+        return download_state_doc(config)
+    except boto3.client("s3").exceptions.NoSuchKey:
+        return None
+
+
+def get_or_create_state_doc_in_s3(config):
+    """get (or create if missing) the state_doc in S3."""
+    state_doc = get_state_doc_in_s3(config)
+    if state_doc is None:
+        state_doc = create_state_doc(config)
+    return state_doc
+
+
+def get_or_create_state_doc_in_path(config):
+    try:
+        with open(config["state_doc_path"], 'r') as json_file:
+            state_doc = json.load(json_file)
+    except IOError:
```

Comment:
Can't IOError also mean permission issues?  Can probably tell that from the error that will come from trying ot create the doc, but not certain.